### PR TITLE
Test that two-dimensional matrix is simd aware.

### LINKED ISSUE #1300: Are matrices simd aware? - check if value type `simd_type` is working in two_dimensional_matrix
**TASK: check if value type `simd_type` is working in two_dimensional_matrix**

- [x] check if `seqan3::detail::two_dimensional_matrix<simd_type<int>>` is working; see [two_dimensional_matrix](https://github.com/seqan/seqan3/blob/8993bb63efa8f844f23bd2207e9bac1061c94bd1/include/seqan3/alignment/matrix/detail/two_dimensional_matrix.hpp#L58-L61) and [simd_type](https://github.com/seqan/seqan3/blob/8993bb63efa8f844f23bd2207e9bac1061c94bd1/include/seqan3/core/simd/simd.hpp#L45-L48)
- [x] add tests to [two_dimensional_matrix_test.cpp](https://github.com/seqan/seqan3/blob/8993bb63efa8f844f23bd2207e9bac1061c94bd1/test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp#L26-L33)
- [x] create some constants `auto one = seqan3::simd::fill<simd_t>(1)`, `auto iota = seqan3::simd::iota<simd_t>()` and fill the matrix with those constants
- [x] compile this test locally with `-msse4` and `-O3` to see if it is working
- [x] (NOTE: Our build system does not support compiling and testing it for different CPU instruction sets, yet)

### PR DESCRIPTION
**Pull request by marehr**
**TASK: check if value type `simd_type` is working in two_dimensional_matrix**

- [x] check if `seqan3::detail::two_dimensional_matrix<simd_type<int>>` is working; see [two_dimensional_matrix](https://github.com/seqan/seqan3/blob/8993bb63efa8f844f23bd2207e9bac1061c94bd1/include/seqan3/alignment/matrix/detail/two_dimensional_matrix.hpp#L58-L61) and [simd_type](https://github.com/seqan/seqan3/blob/8993bb63efa8f844f23bd2207e9bac1061c94bd1/include/seqan3/core/simd/simd.hpp#L45-L48)
- [x] add tests to [two_dimensional_matrix_test.cpp](https://github.com/seqan/seqan3/blob/8993bb63efa8f844f23bd2207e9bac1061c94bd1/test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp#L26-L33)
- [x] create some constants `auto one = seqan3::simd::fill<simd_t>(1)`, `auto iota = seqan3::simd::iota<simd_t>()` and fill the matrix with those constants
- [x] compile this test locally with `-msse4` and `-O3` to see if it is working
- [x] (NOTE: Our build system does not support compiling and testing it for different CPU instruction sets, yet)

----

Fixes #1300 .

### CODE REVIEW

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp [hunk 1]
```diff
@@ -8,31 +8,109 @@
...
+struct two_dimensional_matrix_test<test_matrix_t<score_t, order>> : public ::testing::Test
+{
+    using matrix_type = test_matrix_t<score_t, order>;
+    using score_type = typename matrix_type::value_type;
+    static constexpr matrix_major_order matrix_order = order;
+
+    std::vector<score_type> expected_matrix_cell{create_matrix_storage<score_type, matrix_major_order::row>()};
+    std::vector<score_type> matrix_storage{create_matrix_storage<score_type, matrix_order>()};
+
+    // note: that we construct the internal data representation of the matrix dependent on the matrix_major_order.
```
Comment (eseiler) [inline test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]:
```suggestion
    // Note: We construct the internal data representation of the matrix dependent on the matrix_major_order.
```

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp [hunk 2]
```diff
@@ -8,31 +8,109 @@
...
+    };
+
+    if constexpr(seqan3::simd_concept<score_type>)
+    {
+        row_wise[5] = seqan3::simd::iota<score_type>(5);
+        row_wise[8] = seqan3::simd::fill<score_type>(8);
+
+        column_wise[4] = seqan3::simd::iota<score_type>(5);
+        column_wise[2] = seqan3::simd::fill<score_type>(8);
+    }
```
Comment (eseiler) [inline test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]: What does this do?

**Comment by marehr** [comment]
Okay let's meet up tomorrow.

File: include/seqan3/alignment/matrix/detail/two_dimensional_matrix.hpp [hunk 3]
```diff
@@ -142,6 +142,48 @@ class two_dimensional_matrix
...
+     * 0, 4, 8, 1, 5, 9, 2, 6, 10, 3, 7, 11
+     *
+     * Note that the matrix is not transposed, so that the general layout as displayed above will remain the same.
+     * It only changes the matrix major order, i.e. data stored row wise is now stored column wise and vice versa.
+     */
+    template <typename other_value_t, typename other_allocator_t, matrix_major_order other_order>
+    //!\cond
+        requires std::assignable_from<other_value_t &, value_t &>
+    //!\endcond
+    explicit constexpr two_dimensional_matrix(two_dimensional_matrix<other_value_t, other_allocator_t, other_order> const & matrix) :
```
Comment (eseiler) [inline include/seqan3/alignment/matrix/detail/two_dimensional_matrix.hpp]: line length

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp (hunk 2 above)
Comment (eseiler) [reply test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]: Add a comment why fill and iota is needed

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp [hunk 4]
```diff
@@ -78,249 +167,183 @@ TYPED_TEST(two_dimensional_matrix_test, rows)
...
+    using matrix_type = test_matrix_t<score_t, order>;
+    using base_t = two_dimensional_matrix_test<matrix_type>;
 
-    // Test random access range.
+    // test_range is a random access range.
     using iterator_tag = std::random_access_iterator_tag;
-    static constexpr bool const_iterable = true;
-
-    matrix_type test_range{number_rows{3}, number_cols{4}, base_t::expected_range};
+    static constexpr bool const_iterable = false;
```
Comment (eseiler) [inline test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]: Look into why this has now to be false.

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp [hunk 5]
```diff
@@ -78,249 +167,183 @@ TYPED_TEST(two_dimensional_matrix_test, rows)
...
+struct two_dimensional_matrix_iterator_test : two_dimensional_matrix_test<matrix_t>
 {
-    using base_t = two_dimensional_matrix_test<policy_t>;
-    using matrix_type = typename base_t::matrix_type;
+    using matrix_type = matrix_t;
+    using base_t = two_dimensional_matrix_test<matrix_type>;
     using iterator_type = typename matrix_type::iterator;
 
-    matrix_type test_range{number_rows{3}, number_cols{4}, base_t::expected_range};
+    matrix_type test_range{number_rows{3}, number_cols{4}, base_t::expected_matrix_cell};
```
Comment (eseiler) [inline test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]: expected matrix content

File: include/seqan3/alignment/matrix/detail/two_dimensional_matrix.hpp [hunk 6]
```diff
@@ -142,6 +142,48 @@ class two_dimensional_matrix
         assert(static_cast<size_t>(std::ranges::distance(entries)) == (row_dim.get() * col_dim.get()));
         storage = std::move(entries);
     }
+
+    /*!\brief Explicit transformation to the other major-order.
```
Comment (eseiler) [inline include/seqan3/alignment/matrix/detail/two_dimensional_matrix.hpp]:
```suggestion
    /*!\brief Explicit construction from the other major-order.
```

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp [hunk 7]
```diff
@@ -8,31 +8,117 @@
...
+struct two_dimensional_matrix_test<test_matrix_t<score_t, order>> : public ::testing::Test
+{
+    using matrix_type = test_matrix_t<score_t, order>;
+    using score_type = typename matrix_type::value_type;
+    static constexpr matrix_major_order matrix_order = order;
+
+    std::vector<score_type> expected_matrix_cell{create_matrix_storage<score_type, matrix_major_order::row>()};
+    std::vector<score_type> matrix_storage{create_matrix_storage<score_type, matrix_order>()};
+
+    // Note: We construct the internal data representation of the matrix dependent on the matrix_major_order.
```
Comment (eseiler) [inline test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]:
```suggestion
    // Note: We construct the internal data representation of the matrix depending on the matrix_major_order.
```

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp [hunk 8]
```diff
@@ -78,249 +167,183 @@ TYPED_TEST(two_dimensional_matrix_test, rows)
...
-    else  // Column order.
-    {
-        EXPECT_EQ((matrix[{row_index_type{0u}, column_index_type{0u}}]), 0);
-        EXPECT_EQ((matrix[{row_index_type{1u}, column_index_type{0u}}]), 1);
-        EXPECT_EQ((matrix[{row_index_type{0u}, column_index_type{1u}}]), 3);
-        EXPECT_EQ((matrix[{row_index_type{0u}, column_index_type{3u}}]), 9);
-        EXPECT_EQ((matrix[{row_index_type{2u}, column_index_type{0u}}]), 2);
-        EXPECT_EQ((matrix[{row_index_type{2u}, column_index_type{3u}}]), 11);
-    }
+    // note: even if the internal storage has a different data layout, accessing the data via a matrix coordinate
```
Comment (eseiler) [inline test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]:
```suggestion
    // Note: Even if the internal storage has a different data layout, accessing the data via a matrix coordinate
```

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp [hunk 9]
```diff
@@ -78,249 +167,183 @@ TYPED_TEST(two_dimensional_matrix_test, rows)
...
-        EXPECT_EQ(converted.cols(), matrix.cols());
-        EXPECT_EQ((converted[{row_index_type{0u}, column_index_type{0u}}]), 0u);
-        EXPECT_EQ((converted[{row_index_type{1u}, column_index_type{0u}}]), 4u);
-        EXPECT_EQ((converted[{row_index_type{0u}, column_index_type{1u}}]), 1u);
-        EXPECT_EQ((converted[{row_index_type{0u}, column_index_type{3u}}]), 3u);
-        EXPECT_EQ((converted[{row_index_type{2u}, column_index_type{0u}}]), 8u);
-        EXPECT_EQ((converted[{row_index_type{2u}, column_index_type{3u}}]), 11u);
-    }
-    else
+    // test that changing the matrix layout works
```
Comment (eseiler) [inline test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]:
```suggestion
    // Test that changing the matrix layout works.
```

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp [hunk 10]
```diff
@@ -78,249 +167,183 @@ TYPED_TEST(two_dimensional_matrix_test, rows)
...
+    static constexpr matrix_major_order other_matrix_order = TestFixture::matrix_order == matrix_major_order::row ?
+                                                             matrix_major_order::column : matrix_major_order::row;
+
+    // test that the implicit conversion of the underlying value_type works
```
Comment (eseiler) [inline test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]:
```suggestion
    // Test that the implicit conversion of the underlying value_type works.
```

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp [hunk 11]
```diff
@@ -78,249 +167,183 @@ TYPED_TEST(two_dimensional_matrix_test, rows)
...
+    // (this does not work for every simd backend, in this case we use the same score type)
```
Comment (eseiler) [inline test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]:
```suggestion
    // This does not work for every SIMD backend, in such a case we use the original score type.
```

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp [hunk 12]
```diff
@@ -78,249 +167,183 @@ TYPED_TEST(two_dimensional_matrix_test, rows)
...
+    using score_type = typename TestFixture::score_type;
+    using unsigned_score_type = typename make_unsigned_score_type<score_type>::type;
+    using new_score_type = std::conditional_t<std::assignable_from<score_type &, unsigned_score_type &>,
+                                              unsigned_score_type,
+                                              score_type>;
+
+    // the new matrix type that has a new matrix major order and might have an implicit value conversion
```
Comment (eseiler) [inline test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]:
```suggestion
    // The new matrix type that has a new matrix major order and might have an implicit value conversion.
```

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp [hunk 13]
```diff
@@ -78,249 +167,183 @@ TYPED_TEST(two_dimensional_matrix_test, rows)
...
+    using converted_matrix_t = test_matrix_t<new_score_type, other_matrix_order>;
+    converted_matrix_t converted_matrix{this->matrix};
+
+    EXPECT_EQ(converted_matrix.rows(), this->matrix.rows());
+    EXPECT_EQ(converted_matrix.cols(), this->matrix.cols());
+
+    // note: we changed the internal data layout, but accessing the converted matrix by a coordinate yields the same
```
Comment (eseiler) [inline test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]:
```suggestion
    // Note: We changed the internal data layout, but accessing the converted matrix by a coordinate yields the same
```

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp [hunk 14]
```diff
@@ -78,249 +167,183 @@ TYPED_TEST(two_dimensional_matrix_test, rows)
...
-    static constexpr bool const_iterable = true;
-
-    matrix_type test_range{number_rows{3}, number_cols{4}, base_t::expected_range};
+    static constexpr bool const_iterable = false;
+
+    // This one is a bit tricky. The one dimensional overloads of the iterator advancing operations, like it+=5, have
+    // the property that they advance the iterator in major matrix order, e.g. it+=5 in a row-wise matrix will be the
+    // same as it+=matrix_offset{row{0}, col{5}} and it+=5 in a column-wise matrix will be the same as
+    // it+=matrix_offset{row{5}, col{0}}. This effectively means that the iterator behaves exactly like the iterator of
+    // the internal data storage. Thus, passing expected_range as data storage into test_range, will result in an equal
```
Comment (eseiler) [inline test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]:
```suggestion
    // the internal data storage, i.e. the iterator's behaviour does not depend on the major matrix order. Thus, passing expected_range as data storage into test_range, will result in an equal
```

adjust line length if suggestion is added

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp [hunk 15]
```diff
@@ -78,249 +172,183 @@ TYPED_TEST(two_dimensional_matrix_test, rows)
...
+    // test_range is a random access range.
     using iterator_tag = std::random_access_iterator_tag;
     static constexpr bool const_iterable = true;
 
-    matrix_type test_range{number_rows{3}, number_cols{4}, base_t::expected_range};
+    // This one is a bit tricky. The one dimensional overloads of the iterator advancing operations, like it+=5, have
+    // the property that they advance the iterator in major matrix order, e.g. it+=5 in a row-wise matrix will be the
+    // same as it+=matrix_offset{row{0}, col{5}} and it+=5 in a column-wise matrix will be the same as
+    // it+=matrix_offset{row{5}, col{0}}. This effectively means that the iterator behaves exactly like the iterator of
+    // the internal data storage, i.e. the iterator's behaviour does not depend on the major matrix order. Thus, passing expected_range as data storage into test_range, will result in an equal
```
Comment (eseiler) [inline test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]: line length

Comment (marehr) [reply test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]: done

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp [hunk 16]
```diff
@@ -78,249 +172,183 @@ TYPED_TEST(two_dimensional_matrix_test, rows)
...
-        EXPECT_EQ((converted[{row_index_type{1u}, column_index_type{0u}}]), 4u);
-        EXPECT_EQ((converted[{row_index_type{0u}, column_index_type{1u}}]), 1u);
-        EXPECT_EQ((converted[{row_index_type{0u}, column_index_type{3u}}]), 3u);
-        EXPECT_EQ((converted[{row_index_type{2u}, column_index_type{0u}}]), 8u);
-        EXPECT_EQ((converted[{row_index_type{2u}, column_index_type{3u}}]), 11u);
-    }
-    else
+    // Test that changing the matrix layout works.
+    static constexpr matrix_major_order other_matrix_order = TestFixture::matrix_order == matrix_major_order::row ?
+                                                             matrix_major_order::column : matrix_major_order::row;
```
Comment (rrahn) [inline test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]: can you please wrap the command like so:
```
cond 
    ? on_true
    : on_false
```

Comment (marehr) [reply test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]: I couldn't find that rule in https://github.com/seqan/seqan3/wiki/General-formatting, but I will change it for you :heart:

**Pull request merged**